flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableTestBase.java

```diff
@@ -27,44 +27,101 @@
 import org.apache.kafka.clients.CommonClientConfigs;
 import org.apache.kafka.clients.admin.AdminClient;
 import org.apache.kafka.clients.admin.NewTopic;
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.clients.admin.TopicListing;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.junit.After;
+import org.junit.AfterClass;
 import org.junit.Before;
-import org.junit.ClassRule;
+import org.junit.BeforeClass;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.testcontainers.containers.KafkaContainer;
 import org.testcontainers.containers.Network;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
 import org.testcontainers.utility.DockerImageName;
 
+import java.time.Duration;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.stream.Collectors;
 
 /** Base class for Kafka Table IT Cases. */
 public abstract class KafkaTableTestBase extends AbstractTestBase {
 
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaTableTestBase.class);
+
     private static final String INTER_CONTAINER_KAFKA_ALIAS = "kafka";
     private static final Network NETWORK = Network.newNetwork();
     private static final int zkTimeoutMills = 30000;
 
-    @ClassRule
-    public static final KafkaContainer KAFKA_CONTAINER =
-            new KafkaContainer(DockerImageName.parse(DockerImageVersions.KAFKA))
-                    .withEmbeddedZookeeper()
-                    .withNetwork(NETWORK)
-                    .withNetworkAliases(INTER_CONTAINER_KAFKA_ALIAS);
+    public static KafkaContainer kafkaContainer;
 
     protected StreamExecutionEnvironment env;
     protected StreamTableEnvironment tEnv;
 
+    // Timer for scheduling logging task if the test hangs
+    private final Timer loggingTimer = new Timer("Debug Logging Timer");
+
+    @BeforeClass
+    public static void startTestEnv() {
+        kafkaContainer =
+                new KafkaContainer(DockerImageName.parse(DockerImageVersions.KAFKA))
+                        .withEmbeddedZookeeper()
+                        .withNetwork(NETWORK)
+                        .withNetworkAliases(INTER_CONTAINER_KAFKA_ALIAS);
+        kafkaContainer.start();
+        if (LOG.isInfoEnabled()) {
+            kafkaContainer.followOutput(new Slf4jLogConsumer(LOG));
+        }
+    }
+
+    @AfterClass
+    public static void tearDownTestEnv() {
+        kafkaContainer.stop();
+    }
```

Comment:
Is it possible to achieve this by overriding `KafkaContainer#doStart()`?
What the current approach make me uncomfortable are:
- It makes `kafkaContainer` non-final, which lead to potential problems like NPEs or so.
- As an abstract class, `startTestEnv` and `tearDownTestEnv` could be overridden.

Reply:
Fixed by d6bfdd9